File: mock-server/src/main/java/com/buyansong/imserver/protocol/ImEnvelopeCodec.java
package com.buyansong.imserver.protocol;

import com.buyansong.im.protocol.v2.ImEnvelope;
import com.google.protobuf.InvalidProtocolBufferException;

/**
 * Validation wrapper around the generated Protobuf {@link ImEnvelope} codec.
 * Mirrors the Android {@code ImEnvelopeCodec}: same size limit, same version
 * check, same payload-presence check, same {@link ProtocolException} surface.
 */
public final class ImEnvelopeCodec {
    public static final int PROTOCOL_VERSION = 2;
    public static final int MAX_ENVELOPE_BYTES = 64 * 1024;

    private ImEnvelopeCodec() {
    }

    public static byte[] encode(ImEnvelope envelope) {
        validate(envelope);
        byte[] bytes = envelope.toByteArray();
        if (bytes.length > MAX_ENVELOPE_BYTES) {
            throw new ProtocolException("Envelope too large: " + bytes.length + " bytes > " + MAX_ENVELOPE_BYTES);
        }
        return bytes;
    }

    public static ImEnvelope decode(byte[] bytes) {
        if (bytes == null || bytes.length == 0) {
            throw new ProtocolException("Empty envelope");
        }
        if (bytes.length > MAX_ENVELOPE_BYTES) {
            throw new ProtocolException("Envelope too large: " + bytes.length + " bytes > " + MAX_ENVELOPE_BYTES);
        }
        ImEnvelope envelope;
        try {
            envelope = ImEnvelope.parseFrom(bytes);
        } catch (InvalidProtocolBufferException error) {
            throw new ProtocolException("Malformed envelope: " + error.getMessage());
        }
        validate(envelope);
        return envelope;
    }

    private static void validate(ImEnvelope envelope) {
        if (envelope.getProtocolVersion() != PROTOCOL_VERSION) {
            throw new ProtocolException("Unsupported protocol version: " + envelope.getProtocolVersion());
        }
        if (envelope.getPayloadCase() == ImEnvelope.PayloadCase.PAYLOAD_NOT_SET) {
            throw new ProtocolException("Envelope payload is not set");
        }
    }
}

File: app/src/main/java/com/buyansong/im/protocol/ImEnvelopeCodec.kt
package com.buyansong.im.protocol

import com.buyansong.im.protocol.v2.ImEnvelope
import com.google.protobuf.InvalidProtocolBufferException

/**
 * Validation wrapper around the generated Protobuf [ImEnvelope] codec.
 *
 * Enforces the wire-format invariants of protocol version 2 (one binary
 * WebSocket message = one envelope): decoded/encoded size limit,
 * `protocol_version == 2`, and a set `oneof payload`. All failures surface
 * as [ProtocolException] so callers keep a single failure path.
 */
object ImEnvelopeCodec {
    const val PROTOCOL_VERSION = 2
    const val MAX_ENVELOPE_BYTES = 64 * 1024

    fun encode(envelope: ImEnvelope): ByteArray {
        validate(envelope)
        val bytes = envelope.toByteArray()
        if (bytes.size > MAX_ENVELOPE_BYTES) {
            throw ProtocolException("Envelope too large: ${bytes.size} bytes > $MAX_ENVELOPE_BYTES")
        }
        return bytes
    }

    fun decode(bytes: ByteArray): ImEnvelope {
        if (bytes.isEmpty()) {
            throw ProtocolException("Empty envelope")
        }
        if (bytes.size > MAX_ENVELOPE_BYTES) {
            throw ProtocolException("Envelope too large: ${bytes.size} bytes > $MAX_ENVELOPE_BYTES")
        }
        val envelope = try {
            ImEnvelope.parseFrom(bytes)
        } catch (error: InvalidProtocolBufferException) {
            throw ProtocolException("Malformed envelope: ${error.message}")
        }
        validate(envelope)
        return envelope
    }

    private fun validate(envelope: ImEnvelope) {
        if (envelope.protocolVersion != PROTOCOL_VERSION) {
            throw ProtocolException("Unsupported protocol version: ${envelope.protocolVersion}")
        }
        if (envelope.payloadCase == ImEnvelope.PayloadCase.PAYLOAD_NOT_SET) {
            throw ProtocolException("Envelope payload is not set")
        }
    }
}
